[FEATURE] Warn users if more specific SQL Datasource is available.

Kilo59 opened the pull request:
Users can almost always use the generic `SQLDatasource` for any SQL dialect, but the experience will be better if using a more specific Datasource.

### Proposed Change

Compare the `connection_string` with our Datasource `type_lookup` to see if a more specific datasource is available and raise a warning if it is.
The warning links to our documentation and suggests a more specific Datasource class.


### Testing Considerations

Added a `filter_gx_datasource_warnings` fixture that can be used to temporarily filter out this warning in tests.

Kilo59 committed:
add `_warn_for_more_specific_datasource_type`

Kilo59 committed:
handle `ConfigStr`

wookasz commented on great_expectations/datasource/fluent/sql_datasource.py:
```diff
@@ -1014,6 +1015,36 @@ def execution_engine_type(self) -> Type[SqlAlchemyExecutionEngine]:
+            # TODO: handle postgresql vs postgres type
+            if more_specific_datasource:
+                warnings.warn(
+                    f"You a using a generic SQLDatasource but a more specific {more_specific_datasource.__name__} "
```
```suggestion
                    f"You are using a generic SQLDatasource but a more specific {more_specific_datasource.__name__} "
```

Kilo59 committed:
don't alter `v`

Kilo59 committed:
spelling

ghost commented:
<details open>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=8914&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations">Click on the image</a> for a new way to code review</b>
</summary>

####

[![Review these changes using an interactive CodeSee Map](https://s3.us-east-2.amazonaws.com/maps.codesee.io/images/github/great-expectations/great_expectations/8914/a6e72a5f/a5bbda68ceb8ad74bf68b1013d43ecace7b332d3.svg)](https://app.codesee.io/r/reviews?pr=8914&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)

#### Legend
<img src="https://app.codesee.io/Legend_v3.png" alt="CodeSee Map legend" width="220" height="156">
</details>

Kilo59 commented on great_expectations/datasource/fluent/sql_datasource.py:
```diff
@@ -1014,6 +1015,40 @@ def execution_engine_type(self) -> Type[SqlAlchemyExecutionEngine]:
+        """
+        Warns if a more specific datasource type may be more appropriate based on the connection string connector prefix.
+        """
+        # TODO: do this when when creating the engine instead of here?
```
If we do this when we create the engine we will already have done the config/env variable substitution so we will be able to warn users that have something like `"${MY_FULL_CONNECTION_STRING}"`.

Kilo59 committed:
`GxDatasourceWarning`

Kilo59 committed:
check for more specific datasource type when creating engine

This will result in warnings everytime we call `.test_connection()` which may be better then `__init__`. It also allow us to check `ConfigStr` connector types.

Kilo59 committed:
simplify

Kilo59 committed:
Merge branch 'develop' into f/_/generic-sql-warning

Kilo59 committed:
test warnings

Kilo59 committed:
no warnings case

Kilo59 committed:
handle `postgresql` + `databricks_sql` connectors

Kilo59 committed:
update `_SourceFactories` type

Kilo59 committed:
test `ConfigStr`

Kilo59 committed:
don't fail if create_engine_spy not called

Kilo59 committed:
add `filter_gx_datasource_warnings` fixture

Kilo59 committed:
remove TODO

Kilo59 committed:
Merge branch 'develop' into f/_/generic-sql-warning

Kilo59 committed:
filter datasource warnings

NathanFarmer commented on tests/datasource/fluent/test_sql_datasources.py:
```diff
@@ -189,5 +210,45 @@ def test_table_quoted_name_type_all_lower_case_normalizion_full():
+@pytest.mark.parametrize(
+    ["connection_string", "suggested_datasource_class"],
+    [
+        ("gregshift://", None),
+        ("sqlite:///", "SqliteDatasource"),
+        ("snowflake+pyodbc://", "SnowflakeDatasource"),
+        ("postgresql+psycopg2://bob:secret@localhost:5432/my_db", "PostgresDatasource"),
+        ("${MY_PG_CONN_STR}", "PostgresDatasource"),
+        ("databricks://", "DatabricksSQLDatasource"),
```
❤️
